selfdrive/controls/lib/longitudinal_planner.py

```diff
@@ -60,158 +58,90 @@ class Planner():
+    accel_limits_turns = limit_accel_in_turns(v_ego, sm['carState'].steeringAngleDeg, accel_limits, self.CP)
+    if force_slow_decel:
+      # if required so, force a smooth deceleration
+      accel_limits_turns[1] = min(accel_limits_turns[1], AWARENESS_DECEL)
+      accel_limits_turns[0] = min(accel_limits_turns[0], accel_limits_turns[1])
+    # clip limits, cannot init MPC outside of bounds
+    accel_limits_turns[0] = min(accel_limits_turns[0], self.a_desired)
+    accel_limits_turns[1] = max(accel_limits_turns[1], self.a_desired)
+    self.mpcs['cruise'].set_accel_limits(accel_limits_turns[0], accel_limits_turns[1])
+
+    next_a = np.inf
+    for key in self.mpcs:
+      self.mpcs[key].set_cur_state(self.v_desired, self.a_desired)
+      self.mpcs[key].update(sm['carState'], sm['modelV2'], v_cruise)
+      if self.mpcs[key].status and self.mpcs[key].mpc_solution.a_ego[5] < next_a:
+        self.longitudinalPlanSource = key
+        self.a_desired_trajectory = list(self.mpcs[key].mpc_solution.a_ego)[:CONTROL_N]
+        next_a = self.mpcs[key].mpc_solution.a_ego[5]
+
+    # Throw FCW if brake predictions exceed capability
+    self.fcw = (bool(np.any(np.array(self.a_desired_trajectory) < -3.0)) and
+                sm['modelV2'].leads[0].prob > .9 and
+                sm['modelV2'].leads[0].prob > .9)
 
     # Interpolate 0.05 seconds and save as starting point for next iteration
-    a_acc_sol = self.a_acc_start + (CP.radarTimeStep / LON_MPC_STEP) * (self.a_acc - self.a_acc_start)
-    v_acc_sol = self.v_acc_start + CP.radarTimeStep * (a_acc_sol + self.a_acc_start) / 2.0
-    self.v_acc_next = v_acc_sol
-    self.a_acc_next = a_acc_sol
+    a_prev = self.a_desired
+    self.a_desired = np.interp(DT_MDL, T_IDXS[:CONTROL_N], self.a_desired_trajectory)
+    self.v_desired = self.v_desired + DT_MDL * (self.a_desired + a_prev)/2.0
 
-    self.first_loop = False
 
   def publish(self, sm, pm):
-    self.mpc1.publish(pm)
-    self.mpc2.publish(pm)
-
     plan_send = messaging.new_message('longitudinalPlan')
 
-    plan_send.valid = sm.all_alive_and_valid(service_list=['carState', 'controlsState', 'radarState'])
+    plan_send.valid = sm.all_alive_and_valid(service_list=['carState', 'controlsState'])
 
     longitudinalPlan = plan_send.longitudinalPlan
     longitudinalPlan.mdMonoTime = sm.logMonoTime['modelV2']
-    longitudinalPlan.radarStateMonoTime = sm.logMonoTime['radarState']
-
-    longitudinalPlan.vCruise = float(self.v_cruise)
-    longitudinalPlan.aCruise = float(self.a_cruise)
-    longitudinalPlan.vStart = float(self.v_acc_start)
-    longitudinalPlan.aStart = float(self.a_acc_start)
-    longitudinalPlan.vTarget = float(self.v_acc)
-    longitudinalPlan.aTarget = float(self.a_acc)
-    longitudinalPlan.vTargetFuture = float(self.v_acc_future)
-    longitudinalPlan.hasLead = self.mpc1.prev_lead_status
+    longitudinalPlan.radarStateMonoTime = 0
+
+    longitudinalPlan.vStart = float(self.v_desired)
```

Comment:
This is using values 0.05s in the future. Think all the future projection should be handled in controlsd

Reply:
Yeah I agree, still need to do that.